### PR TITLE
Remove f parameter

### LINKED ISSUE #36: Decide what to do with `Claim f` parameter
Following #33 , we 'temporarily' removed the only other implementation of `f`, pure functions in `Daml.ContingentClaims.ObservableFunctions.daml`.

This was useful for demos, and is a good mental model to indicate that `f` behaves like a `* -> *`. However in  practice it's not so useful as `Claim (->) * *` can't be serialized (i.e. functions can't).

We need to decide whether we should remove the `f` type parameter - the advantage being simpler signatures I suppose.

The downside is losing some structure that we get from keeping `f` abstract, and the inability for simple demos like stated above.

### PR DESCRIPTION
**Pull request by lucianojoublanc-da**
Fix #36 , by removing the `f` parameter to simplify the code.

This also changes the representation of `And` and `Or` to have multiple children (rather than two). It's isomorphic to the previous representation, as you can always fold/unfold nested `And`s/`Or`s into a list, assuming you have at least two elements.
Prior to this we had a separate `Claim.Serializable` type which represented the on-ledger format, which actually did have the flattened version of `And`. But this was clunky as we always had to deserialize before operating on the types and then do the converse when writing to the ledger. 
The original requirement is described in issue #7 and #25.

### CODE REVIEW

File: QUICKSTART.md [hunk 1]
```diff
@@ -66,7 +66,7 @@ zcb maturity principal asset =
 Here we've just wrapped our expression from the previous section in a function `zcb`, that we can reuse to build the `fixed`-rate bond:
 
 ```Haskell
-fixed : Decimal -> Decimal -> a -> [t] -> Claim a
+fixed : Decimal -> Decimal -> a -> [Date] -> Claim a
```
Comment (lucianojoublanc-da) [inline QUICKSTART.md:69]: Not sure why I changed this. Review.

File: daml/ContingentClaims/Claim.daml [hunk 2]
```diff
@@ -72,109 +71,90 @@ mapParams ft' ft fk fv =
...
+  sequence (AndF []) = error "Traversable ClaimF : sequence empty AndF"
+  sequence (OrF [fa]) = (\a -> OrF [a]) <$> fa
+  sequence (OrF (fa :: fas)) = or <$> fa <*> Prelude.sequence fas
+    where or a as = OrF (a :: as)
+  sequence (OrF []) = error "Traversable ClaimF : sequence empty OrF"
   sequence (CondF p fa fa') = CondF p <$> fa <*> fa'
   sequence (AnytimeF p fa) = AnytimeF p <$> fa
   sequence (UntilF p fa) = UntilF p <$> fa
+
+-- Inequality --
```
Comment (lucianojoublanc-da) [inline daml/ContingentClaims/Claim.daml:142]: Should I put this in it's own file?

File: daml/ContingentClaims/MathML.daml [hunk 3]
```diff
@@ -54,12 +54,13 @@ instance ToXml t => ToXml (Expr t) where
     SupF t τ x -> (mo "Sup" `msub` mfenced "{" "}" bound ) :: mo "&ApplyFunction;" :: x.attribute
                   where bound = [presentation t, mo "&le;", presentation τ]
     IdentF t -> [presentation t]
-    AddF x (Cofree _ (NegF x')) -> x.attribute ++ [mo "-"] ++ x'.attribute
-    AddF x x' -> x.attribute ++ [mo "+"] ++ x'.attribute
+    SumF xs -> foldr add [] xs
```
Comment (lucianojoublanc-da) [inline daml/ContingentClaims/MathML.daml:57]: I should consider adding some more pattern matches here to output terms using big-sigma notation when possible. For instance, if I have a coupon paying bond, it would be nice to be able to output something like  `\sum_{i = Dec, Aug}{When i ...}`, instead of the individual terms.

File: QUICKSTART.md (hunk 1 above)
Comment (lucianojoublanc-da) [reply QUICKSTART.md:69]: This is in the documentation so doesn't matter.

File: daml/ContingentClaims/Math/Stochastic.daml [hunk 4]
```diff
@@ -136,16 +138,19 @@ fapf ccy disc exch today = flip evalState 0 . futuM coalg . Left . (, today) whe
   ϵ (One asset, t) = pure $ exch' asset ccy t
   ϵ (Give c, t) = pure . NegF . claim $ (c, t)
   ϵ (Scale k c, t) = pure $ obs (k, t) `MulF` claim (c, t)
-  ϵ (And c c', t) = pure $ claim (c, t) `AddF` claim (c', t)
-  ϵ (Or c c', t) = pure $ AddF
-      (ind (claim (c , t)) (claim (c', t)) * claim (c , t))
+  ϵ (And cs, t) = pure . SumF $ fmap (claim . (, t)) cs 
+  ϵ (Or [c, c'], t) = pure $ SumF [
```
Comment (lucianojoublanc-da) [inline daml/ContingentClaims/Math/Stochastic.daml:142]: Need to handle multiple-element lists.

This is achieved by permuting the indicator function as if it were a 'binary' number.
Possibly consider using a combinatorial representation here e.g. `nCr` here instead of expanding into a huge sum of indicator functions, which will quickly grow large.

**Comment by lucianojoublanc-da** [comment]
I just thought now (after that refactor :grimacing: ) that I should declare the branching nodes as e.g.

```haskell
data Claim = And { first : Claim, second : Claim, rest : [Claim] }
```

This would ensure that they have **at least** 2 elements, by construction. It would even be enforced off-ledger, when working in other languages. But it is a bit weird for someone who's not familiar with this. It's the same as `DA.NonEmpty` if you've ever used that.

File: daml/ContingentClaims/Claim.daml (hunk 2 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:142]: This would probably make a nicer structure, but the fact is that `Inequality` is only used in the creation of a `Claim`, so this would just be one more thing people would need to know. In case of a `Claim` blanket import, you don't need to even be aware of this.

File: test/daml/Test/Lifecycle.daml [hunk 5]
```diff
@@ -98,6 +98,8 @@ testSettle = script do
           . apoM (Lifecycle.lifecycle' observe25)
           . (1.0, )
 
+  let and x y = And x y []
```
Comment (lucianojoublanc-da) [inline test/daml/Test/Lifecycle.daml:101]: The new 'at-least 2' `And` datatype added in 28429cd makes it impossible to write 

```
x `And` y 
```
which is why I added this helper function. That is pretty annoying. Not sure if it should be part of the library. We could also consider re-introducing lower-case constructors like we had way back when i.e. 

```haskell
one : Claim
when : Inequality -> Claim -> Claim
and : Claim -> Claim -> Claim
```
and make `and` flatten nested `And`s as you go along.
On one hand that would affect all dependent applications. On the other hand, it would maintain the illusion of having nested `And` nodes. i.e. you would still write
```
a `and` b `and` c
```
in other words, it would just require lowercasing all combinators.

File: test/daml/Test/Lifecycle.daml [hunk 6]
```diff
@@ -110,12 +112,12 @@ testSettle = script do
   remaining === Give Zero
   pending === [(-1.0 , a)]
 
-  Lifecycle.Result{pending, remaining} <- f (Scale two (And [One a, One b])) today
-  remaining === Scale two  (And [Zero, Zero])
```
Comment (lucianojoublanc-da) [inline test/daml/Test/Lifecycle.daml]: Note that this is much clunkier than the 
```
a `And` b
```
notation

File: test/daml/Test/Lifecycle.daml (hunk 5 above)
Comment (lucianojoublanc-da) [reply test/daml/Test/Lifecycle.daml:101]: A downside we previously had with these constructors was the clash with `Prelude.and`. That was pretty annoying too.

File: daml/ContingentClaims/Claim.daml [hunk 7]
```diff
@@ -23,45 +22,45 @@ type F = ClaimF
...
+  | Give (Claim t x a)
       -- ^ The obligations of the bearer and issuer are revesed.
-  | And with lhs: Claim f t x a, rhs: Claim f t x a
+  | And with fst: Claim t x a, snd: Claim t x a, tail: [Claim t x a]
       -- ^ Used to combine multiple rights together.
-  | Or with lhs: Claim f t x a, rhs: Claim f t x a
-      -- ^ Gives the bearer the right to choose between two claims.
-  | Cond with predicate: Inequality f t x a, success: Claim f t x a, failure: Claim f t x a --TODO rename predicate -> `lte` or `inequality`
+  | Or with claims: [Claim t x a]
+      -- ^ Gives the bearer the right to choose between several claims.
```
Comment (GeorgSchneider) [inline daml/ContingentClaims/Claim.daml:37]: Shouldn't we have at minimum 2 elements here as well if we have it for `And`?

File: daml/ContingentClaims/Claim.daml [hunk 8]
```diff
@@ -23,45 +22,45 @@ type F = ClaimF
...
+data Claim t x a
   = Zero
       -- ^ Represents an absence of claims.
   | One a
       -- ^ The bearer acquires one unit of `a` *immediately*.
-  | Give (Claim f t x a)
+  | Give (Claim t x a)
       -- ^ The obligations of the bearer and issuer are revesed.
-  | And with lhs: Claim f t x a, rhs: Claim f t x a
+  | And with fst: Claim t x a, snd: Claim t x a, tail: [Claim t x a]
```
Comment (GeorgSchneider) [inline daml/ContingentClaims/Claim.daml:34]: Thinking about the usability of this representation, I'm not entirely convinced this is the best way to enforce this. it prevents us from folding or mapping over `And`s and makes it clunkier to work with it in general. Could we not add a validation function that checks the minimum elements? I also think we could allow `And` with a single element, and it would simplify to just this element instead. And an `And` with zero elements could essentially be pruned off the tree. Just some thoughts, not sure what the best options is here, but worth discussing.

**Review by GeorgSchneider — APPROVED** [review]
Some comments to discuss, but good to merge from my point

File: daml/ContingentClaims/Claim.daml (hunk 8 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:34]: I agree with you that this doesn't feel very nice.

>  it prevents us from folding or mapping over Ands 

That's not quite right, unless I've misunderstood what you're saying:
```
instance Functor Claim where
  fmap f (And fst snd tail) = And (f fst) (f snd) (fmap f tail)
```
and ditto for `fold`

> Could we not add a validation function that checks the minimum elements?

The problem is we can't use `ensure` because a `Claim` is not a `template`. So there is no way to enforce it at the data type level.

I mentioned using "smart constructors". I'll look into that again.
Another idea I had was adding a `monoid` instance, so that you can just `And` together multiple claims using the `<>` operator.

File: daml/ContingentClaims/Claim.daml (hunk 7 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:37]: Yes. I was testing out the `And` before doing the same for `Or`.

I did have some doubts whether the binary representation should be kept for `Or`, see my comments above on the bermudan option.

File: daml/ContingentClaims/Financial.daml [hunk 9]
```diff
@@ -20,51 +21,47 @@ unrollDates issueYear maturityYear fixingMonths fixingDay =
...
--- | A floating rate bond. The first two arguments are `Observable`s.
-floating : Observable f t x ccy => f t x ccy -> f t x ccy -> ccy -> [t] -> Claim f t x ccy
-floating principal coupon asset = apo \case
-     [maturity] -> Left (forward maturity coupon (One asset)) `AndF`
-                   Left (forward maturity principal (One asset))
-     (t :: ts) -> Left (forward t coupon (One asset)) `AndF` Right ts
-     [] -> ZeroF
+-- | A floating rate bond.
+floating : forall t x ccy . Observation t x ccy -> Observation t x ccy -> ccy -> [t] -> Claim t x ccy
+floating principal coupon asset fixings = foldMap mkCoupon fixings <> mkPrincipal (last fixings : Optional t) -- TODO: not very efficient, but easy to understand
```
Comment (lucianojoublanc-da) [inline daml/ContingentClaims/Financial.daml:37]: @GeorgSchneider have a look at this - after adding a monoid instance i.e. `(<>)`, we can fold over claims quite nicely, like this implementation shows.

File: daml/ContingentClaims/Claim.daml (hunk 8 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:34]: Re: smart constructors, running into a problem we had last time, raised a bug for it now: https://github.com/digital-asset/daml/issues/12166

Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:34]: Also, I'm consider whether this alternative representation would be appropriate, since we're already messing around with this:

```
And inits : [ Claim] , penultimate : Claim, last : Claim
```

I _think_ this might be considerably more efficient than the current approach. Need to mull over it a bit more.

**Comment by lucianojoublanc-da** [comment]
Merged as part of #45
